src/twisted/web/template.py

```diff
@@ -109,40 +126,43 @@ class _ToStan(handler.ContentHandler, handler.EntityResolver):
     Document Object Model.
     """
 
-    def __init__(self, sourceFilename):
+    def __init__(self, sourceFilename: str):
         """
         @param sourceFilename: the filename to load the XML out of.
         """
         self.sourceFilename = sourceFilename
         self.prefixMap = _NSContext()
         self.inCDATA = False
 
-    def setDocumentLocator(self, locator):
+    def setDocumentLocator(self, locator: Locator) -> None:
         """
         Set the document locator, which knows about line and character numbers.
         """
         self.locator = locator
 
-    def startDocument(self):
+    def startDocument(self) -> None:
         """
         Initialise the document.
         """
-        self.document = []
+        # Depending on our active context, the element type can be Tag, slot
+        # or str. Since mypy doesn't understand that context, it would be
+        # a pain to not use Any here.
+        self.document: List[Any] = []
```

Comment:
"Leave as future exercise" seems totally reasonable in this PR

Reply:
I am not strongly recommending this, just mentioning it, but something that can sometimes be useful in this type of situation is adding a top-level type alias and commenting there, rather than at the call site; so like at module scope doing `TagSoupTerminal = Any` (or a better / more specific name if you can think of one) and then eventually replace it with `TagSoupTerminal = Union[Tag, slot, str]` (or similar) without changing the implementation logic

Reply:
> "Leave as future exercise" seems totally reasonable in this PR

I think this is a case where annotations alone aren't going to be sufficient; it will probably require some restructuring of the parser to make it more mypy-friendly.

> I am not strongly recommending this, just mentioning it, but something that can sometimes be useful in this type of situation is adding a top-level type alias and commenting there, rather than at the call site; so like at module scope doing `TagSoupTerminal = Any` (or a better / more specific name if you can think of one) and then eventually replace it with `TagSoupTerminal = Union[Tag, slot, str]` (or similar) without changing the implementation logic

That sounds like a good strategy in other places, but here it wouldn't save much, since there are only two annotations with `Any` and they're both in the block directly below this comment. If this were a public class, there might be some value in assigning a name for documentation purposes (*), but it's a private class instead.

(*) I think support for documented aliases should be added to pydoctor. That's useful for type aliases in particular; for other objects aliases are typically used to export an object under a different name and in that case the documentation of the original can be moved.

Reply:
Agreed with both ;-)